### DISCUSSION TITLE
Expanded Wavelength Range

### DISCUSSION THREAD

**Discussion by BenLiesVP** [rafael-fuente/Diffraction-Simulations--Angular-Spectrum-Method#3] [Ideas]
Is there a possibility to expand the wavelength range to accommodate typical lasers (ex: Yag, CO2, etc.)?

**Comment by rafael-fuente** [comment]
Yes, I updated the package as well as the repository to accommodate them. Initially, I made only the plot function reproducing the human eye's response to light, so it you try a wavelength outside the visible spectrum you wouldn't see anything. 
I added a new plot function to plot the physical intensity of the diffracted field regardless of human eye's response: `plot_intensity`

This is how to use it for a laser with a wavelength of 1064 nm:

Note: You must reinstall the package in order to use it.

```python
import diffractsim
diffractsim.set_backend("CPU") #Change the string to "CUDA" to use GPU acceleration

from diffractsim import MonochromaticField, ApertureFromImage, mm, nm, cm, W, m


F = MonochromaticField(
    wavelength=1064 * nm, extent_x=15 * mm, extent_y=51 * mm, Nx=1028, Ny=1028, intensity  =  0.1 *W / (m**2)
)

F.add(ApertureFromImage( "./apertures/hexagon.jpg", image_size=(7.6* mm, 7.6 * mm), simulation = F))

F.propagate(80*cm)
I = F.get_intensity()

F.plot_intensity(I, xlim=[-9* mm, 9* mm], ylim=[-9* mm, 9* mm])

# plot square root of field intensity (for improved contrast)
F.plot_intensity(I, xlim=[-9* mm, 9* mm], ylim=[-9* mm, 9* mm], square_root = True)
```

And this is the output:

![Figure_I](https://user-images.githubusercontent.com/37953893/138725706-11ec4303-b82a-459d-b499-0ecd639d84f6.png)

![Figure_root_I](https://user-images.githubusercontent.com/37953893/138725718-c47f542c-1c55-4444-bd32-2c3b79d6b070.png)

**Reply by BenLiesVP** [reply]
I am experiencing errors when using wavelength from a CO2 laser (10,600 nm). 

> monochromatic_simulator.py:220: RuntimeWarning: invalid value encountered in sqrt
  kz = bd.sqrt((2 * bd.pi / self.λ) ** 2 - kx ** 2 - ky ** 2)

Is this something that can be worked around?

**Reply by rafael-fuente** [reply]
Yes, the problem is that I didn't include evanescent modes in the original implementation, and in your CO2 laser example, they are triggered. It's now implemented, reinstall the package and your example will work 🙂

**Reply by BenLiesVP** [reply]
That seemed to do the trick. One last question. What would be the best way to depict a concave mirror? Would this act like a lens in this system where the focal lengths are the same?

**Reply by rafael-fuente** [reply]
What is the setup, and how the concave mirror is placed?
Assuming the paraxial approximation, no losses, and considering the next distance propagated as backward, then yes, except for a constant pi phase factor over the entire wavefront, so you can model it just by inverting the sign of the field.

For example, a concave mirror of R = 160 cm  would be modeled by using:
```python
F.add(Lens(f = 80*cm))
F.E = -F.E
```
